cpp/include/cudf/detail/contiguous_split.hpp

```diff
@@ -44,5 +44,70 @@ packed_columns pack(cudf::table_view const& input,
                     rmm::cuda_stream_view stream,
                     rmm::mr::device_memory_resource* mr);
 
+// opaque implementation of `metadata_builder` since it needs to use
+// `serialized_column`, which is only defined in pack.cpp
+class metadata_builder_impl;
+
+/**
+ * @brief Helper class that creates packed column metadata.
+ *
+ * This class is an interface to the opaque metadata that is used to
+ * describe `contiguous_split` and `pack` results.
+ */
+class metadata_builder {
+ public:
+  /**
+   * @brief Construct a new metadata_builder.
+   *
+   * @param num_root_columns is the number of top-level columns
+   */
+  explicit metadata_builder(size_type num_root_columns);
+
+  /**
+   * @brief Destructor that will be implemented as default, required because metadata_builder_impl
+   * is incomplete at this stage.
+   */
+  ~metadata_builder();
+
+  /**
+   * @brief Add a column to this metadata builder.
+   *
+   * Callers must call this function for the parent column and followed by any children,
+   * in the order maintained in the column/column_view.
+   *
+   * Example: given a table with a nested column "a" with 2 children, and a non-nested column "b":
```

Comment:
Good question. This metadata is really just saying that the column of cudf type string, but not specifying how exactly it is going to be encoded. `contiguous_split` special cases a string column and handles these underlying buffers.

For a list or struct types, in addition to the parent type and metadata, other children information is carried (children types, nullability, number of rows, ..)  these are first-class columns (like strings), so that's the level of detail in the metadata.